The Basics of Machine Learning without Math (#4)
csz-academy/csz.academy: _posts/2022-01-07-The-Basics-of-Machine-Learning-without-math/2022-01-07-The-Basics-of-Machine-Learning-without-math.md

---
layout: post

> title:  "The Basics of Machine Learning without math"

Reviewer: ```suggestion
title:  "The Basics of Machine Learning without Math"
```

Titles are all-uppercase except words like "the", "of", "without"

date:   2022-01-07 19:58:49 +0100

> categories: machine learning

Reviewer: Check whether this gets into two categories or it's used as only one. I don't know enough about jekyll.

Reviewer 2: Will be two categories `/machine/learning/`

Reviewer: In that case it probably needs to be enclosed in  `"`, correct? I think it's confusing that this attribute is called `categories`.

Reviewer 2: I think that won't work out since categories will be part of the URI. I think it should he `"machine-learning"` in that case.

---

## Table of Contents
1. How is a neuronal network structured
2. What is a neuron
3. What is Backpropagation

> 4. How can we use this knowledge in practice

Reviewer: Same thing about upper case lettering applies here, too.

## How is a neuronal network structured

> For many people a neuronal network is magic. If you never learnt what happens behind the scenes it looks like you just throw things in and eventually it can solve a problem <br>

Reviewer: ```suggestion
For many people a neuronal network is magic. If you never learnt what happens behind the scenes it looks like you just throw things in and eventually it can solve a problem.

```

Typically, an empty line creates a new paragraph in markdown. This avoids HTML-cluttered markdown (This is an IMHO).

A neuronal Network, as the name already suggests, is a network of mutliple neurons.

<img src="1200px-Neural_network.svg.png" width="600"><br>

> https://upload.wikimedia.org/wikipedia/commons/thumb/3/3d/Neural_network.svg/1200px-Neural_network.svg.png <br><br>

Reviewer: ```suggestion
<figure>
  <img src="1200px-Neural_network.svg.png" width="600" alt="Simplified representation of a Neural Network with two green input nodes, five blue intermediate nodes and one yellow output node." />
  <figcaption>
    A simplified representation of a neural network with two layers.
    <a href="https://commons.wikimedia.org/wiki/File:Neural_network.svg">(by Dake, Mysid. CC-BY 1.0)</a></figcaption>
</figure>
```

`alt` attributes are important for a11y, and `<figure>`/`<figcaption>` are HTML5 tags for correct grouping of figures and captions. 

License requires attribution.

> You give a input into it and through the interaction of the different neurons the result is formed

Reviewer: ```suggestion
You give a input into it and through the interaction of the different neurons the result is formed.
```

## What is a Neuron

> As you could see in the image above there are 3 types of neurons.

Reviewer: ```suggestion
As you could see in the image above there are 3 types of neurons:
```

And make the next three rows a list - one entry for each neuron type, so:

  - Green
  - Blue
  - Yellow

This would also avoid `<br>` again.

The green ones are input Neurons. The just have connections to their right and no other connections. These are used to input data into the neuronal network. <br>

> The second type are the blue ones. How many collumns of them exist is not given and can be changed. They have connections to their left (input) and to their right (output)

Reviewer: ```suggestion
The second type are the blue ones. How many columns (often called "layers") of them exist is not given and can be changed. They have connections to their left (input) and to their right (output)
```

They take the inputs, process them and then generate an output. <br>

> The last type of neurons are the yellow one. They are at the end of the network. Their Output is the output of the whole network. <br>

Reviewer: ```suggestion
The last type of neurons are the yellow one. They are at the end of the network. Their output is the output of the whole network. <br>
```

All 3 types of the network can be descriped through one model

<img src="1*O7YSSqlOdQuNgMrH7J_3dg.png" width="600"><br>

> https://miro.medium.com/max/428/1*O7YSSqlOdQuNgMrH7J_3dg.png<br><br>

Reviewer: What is the license of this image? Source *seems* to be https://medium.com/analytics-vidhya/neural-network-part1-inside-a-single-neuron-fee5e44f1e

> A neuron has multiple inputs. The inputs are the outputs of all neurons in the previous collum.

Reviewer: ```suggestion
A neuron has multiple inputs. The inputs are the outputs of all neurons in the previous colum.
```

> These inputs get weights. For example if a input is 0.6 and the weight is 0.1 we take 0.6 * 0.1 and then we have the "real" input. Every input has its own weight. All of the inputs multiplied with the weigth are added together. This sum is now put into a function. The function determines how much output is generated for the given sum. Often the function looks like this:

Reviewer: ```suggestion
These inputs are weighted. For example if a input is 0.6 and the weight is 0.1 we take 0.6 * 0.1 and then we have the "real" input. Every input has its own weight. All of the inputs multiplied with the weights are added together. This sum is now put into a function. The function determines how much output is generated for the given sum. Often the function looks like this:
```

Is it possible with jekyll to use the latex-style `$0.6$` or `\(0.6\)` to get nicer typesetting of numbers?

Also - is this the only possible function? (no). Do they have a specific name? (Yes, activation function). What is their purpose? (To add nonlinearity and sometimes to slightly bound the values neurons can output)

Reviewer 2: > Is it possible with jekyll to use the latex-style $0.6$ or \(0.6\) to get nicer typesetting of numbers?

We could add that, sure. I think that's also in your interest.

<img src="1*oePAhrm74RNnNEolprmTaQ.png" width="600"><br>

> https://miro.medium.com/max/357/1*oePAhrm74RNnNEolprmTaQ.png<br><br>

Reviewer: Same problem here with the license. Source seems to be https://medium.com/analytics-vidhya/neural-network-part1-inside-a-single-neuron-fee5e44f1e again

> The output of this function (in this case RELU) is the output of the neuron and therefore the input for the next collum of neurons.<br>

Reviewer: ```suggestion
The output of this function (in this case RELU) is the output of the neuron and therefore the input for the next colum of neurons.<br>
```

What does ReLU mean btw?

> If we have a description of the arrangement of the neurons and a list of all weights we can reproduce the network over and over with the same outputs for the same inputs. And these two informations are the core of a neuronal network. How the Neurons are arrenged and what the weights are. 

Reviewer: ```suggestion
If we have a description of the arrangement of the neurons and a list of all weights we can reproduce the network over and over with the same outputs for the same inputs. And these two informations are the core of a neuronal network: How the Neurons are arranged and what the weights are. 
```

## What is Backpropagation

> The structure of a network impacts its perfomance but the best variable to shape it are the weights. If we get the right weights its should be perfect. Obviously getting good weights is hard. I know i said this is without math but we need a little bit. A neuronal network is like a function with millions of parameters(weights) which produces a deterministic result. We can imagen this like a 3D map. 

Reviewer: ```suggestion
The structure of a network impacts its performance, but the best variable to shape it are the weights. If we get the right weights it should be perfect. Obviously getting good weights is hard. I know that I said this is without math, but we need a little bit. A neuronal network is like a function with millions of parameters(weights) which produces a deterministic result. We can imagine this like a 3D map. 
```

Actually, this is not bound to be 3D, but can also be 2D or significantly higher. 3D is only the case for two input values (x, y axis) and one output (z axis).

I think it's also really important to stress that it's incredibly hard to express that function in a simple and closed form (as it's nonlinear and may express slightly chaotic behavior - a good example if you have a better understanding of NNs might be "Person detection" vs uncanny Valley, but that does not fit well in this post).

Reviewer: Had this idea later on: Maybe introduce error beforehand, so that we can focus on the map as representation of the weights (which you do for gradient descend).

<img src="gradient-descent-local-minima.png" width="600"><br>

> https://bdtechtalks.com/wp-content/uploads/2020/04/gradient-descent-local-minima.png <br><br>

Reviewer: Again, what's the license? Didn't bother to look up the source here.

> The lower we get the better our network performs. Finding the lowest point overall (Global Minima) is at the moment impossible. We take a other approach. We start at a random point somewhere. So we just use random weights. Luckily we have a way to find the way which leads us from our point. This is called negative gradiant. If you are intereseted in math you can research this on your own. <br>

Reviewer: ```suggestion
The lower we get the better our network performs. Finding the lowest point overall (Global Minima) is generally hard, as it is hard to express the desired function at all. We take a other approach. We start at a random point somewhere, so we just use random weights. Luckily we have a way to find the way which leads us from our point. This is called negative gradient. If you are interested in math you can research this on your own. <br>
```

Why do we need to get lower? You likely need to introduce some notion of "Error" for this, and that we can measure this error. And *then* we can say that we want to minimize the error.

Otherwise - why should I care about getting to the lowest point of a function?

> Now lets start learning. We guess some weights and start throwing inputs into the network. We compare the result of the network with the right results. Then we use the negative gradiant and adjust the weights with these metrics. This process is repeated thousand or even millions of times until we find a local minima. If we reach it we cant get out and have to take a look how our network performs. If the accurecy of the network is enough we are finished. When the result are not good enough we have to start again with other starting weights and let the network learn.

Reviewer: ```suggestion
Now lets start learning. We guess some weights and start throwing inputs into the network. We compare the result of the network with the right results. Then we use the negative gradient and adjust the weights with these metrics. This process is repeated thousand or even millions of times until we find a local minima. If we reach it, we can't get out of the region of this minimum and have to take a look how our network performs. If the accuracy of the network is good enough we are finished. When the result is not good enough, we have to start again with other starting weights and repeat the learning process.
```
